generative/maisi/scripts/diff_model_train.py

```diff
@@ -0,0 +1,292 @@
+
+            _iter += 1
+            images = train_data["image"].to(device)
+            images = images * scale_factor
+
+            top_region_index_tensor = train_data["top_region_index"].to(device)
+            bottom_region_index_tensor = train_data["bottom_region_index"].to(device)
+            spacing_tensor = train_data["spacing"].to(device)
+
+            optimizer.zero_grad(set_to_none=True)
+
+            with autocast(enabled=True):
+                noise = torch.randn(
+                    num_images_per_batch,
+                    4,
+                    images.size(-3),
+                    images.size(-2),
+                    images.size(-1),
+                ).to(device)
+
+                timesteps = torch.randint(
+                    0,
+                    num_train_timesteps,
+                    (images.shape[0],),
+                    device=images.device,
+                ).long()
+
+                noisy_latent = noise_scheduler.add_noise(original_samples=images, noise=noise, timesteps=timesteps)
+
+                noise_pred = unet(
+                    x=noisy_latent,
+                    timesteps=timesteps,
+                    top_region_index_tensor=top_region_index_tensor,
+                    bottom_region_index_tensor=bottom_region_index_tensor,
+                    spacing_tensor=spacing_tensor,
+                )
+
+                loss = loss_pt(noise_pred.float(), noise.float())
+
+            scaler.scale(loss).backward()
+            scaler.step(optimizer)
+            scaler.update()
+
+            lr_scheduler.step()
+
+            loss_torch[0] += loss.item()
+            loss_torch[1] += 1.0
+
+            if local_rank == 0:
+                print(
+                    f"[{str(datetime.now())[:19]}] epoch {epoch + 1}, iter {_iter}/{len(train_loader)}, loss: {loss.item():.4f}, lr: {current_lr:.12f}."
+                )
+
+        if torch.cuda.device_count() > 1:
+            dist.all_reduce(loss_torch, op=torch.distributed.ReduceOp.SUM)
+
+        loss_torch = loss_torch.tolist()
+        if torch.cuda.device_count() == 1 or local_rank == 0:
+            loss_torch_epoch = loss_torch[0] / loss_torch[1]
+            print(f"epoch {epoch + 1} average loss: {loss_torch_epoch:.4f}.")
+
+        if local_rank == 0:
+            unet_state_dict = unet.module.state_dict()
+
+            torch.save(
+                {
+                    "epoch": epoch + 1,
+                    "loss": loss_torch_epoch,
+                    "num_train_timesteps": num_train_timesteps,
+                    "scale_factor": scale_factor,
+                    "unet_state_dict": unet_state_dict,
+                },
+                f"{ckpt_folder}/{args.model_filename}",
+            )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Diffusion Model Training")
+    parser.add_argument("--env_config", type=str, required=True, help="Path to environment configuration file")
```

Comment:
Same as the comment above.

Reply:
Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.